pkg/mixer/mixer.go

```diff
@@ -15,168 +15,146 @@
 package mixer
 
 import (
-	"container/list"
 	"sync"
 	"time"
 
 	"github.com/frostbyte73/core"
 
+	"github.com/livekit/sip/pkg/internal/ringbuf"
 	"github.com/livekit/sip/pkg/media"
 )
 
-const (
-	defaultBufferSize = 5 // Samples required for a input before we start mixing
-
-	mixerTickDuration = time.Millisecond * 20 // How ofter we generate mixes
-)
+// inputBufferFrames sets max number of frames that each mixer input will allow.
+// Sending more frames to the input will cause old one to be dropped.
+const inputBufferFrames = 5
 
 type Input struct {
-	mu               sync.Mutex
-	samples          list.List // linked list of elements of type [mixSize]byte
-	ignoredLastPrune bool
-
-	hasBuffered bool
-	bufferSize  int
-	mixSize     int
+	mu  sync.Mutex
+	buf *ringbuf.Buffer[int16]
 }
 
 type Mixer struct {
-	mu sync.Mutex
+	out media.Writer[media.PCM16Sample]
 
-	out    media.Writer[media.PCM16Sample]
-	inputs map[*Input]struct{}
+	mu     sync.Mutex
+	inputs []*Input
 
-	ticker  *time.Ticker
-	mixSize int
+	ticker *time.Ticker
+	mixBuf []int32           // mix result buffer
+	mixTmp media.PCM16Sample // temp buffer for reading input buffers
 
 	stopped core.Fuse
 }
 
-func NewMixer(out media.Writer[media.PCM16Sample], sampleRate int) *Mixer {
-	m := createMixer(out, sampleRate)
+func NewMixer(out media.Writer[media.PCM16Sample], bufferDur time.Duration, sampleRate int) *Mixer {
+	mixSize := int(time.Duration(sampleRate) * bufferDur / time.Second)
+	m := newMixer(out, mixSize)
+	m.ticker = time.NewTicker(bufferDur)
 
 	go m.start()
 
 	return m
 }
 
-func createMixer(out media.Writer[media.PCM16Sample], sampleRate int) *Mixer {
-	m := &Mixer{
+func newMixer(out media.Writer[media.PCM16Sample], mixSize int) *Mixer {
+	return &Mixer{
 		out:     out,
-		ticker:  time.NewTicker(mixerTickDuration),
-		mixSize: int(time.Duration(sampleRate) * mixerTickDuration / time.Second),
+		mixBuf:  make([]int32, mixSize),
+		mixTmp:  make(media.PCM16Sample, mixSize),
 		stopped: core.NewFuse(),
-		inputs:  make(map[*Input]struct{}),
 	}
-
-	return m
 }
 
-func (m *Mixer) doMix() {
-	mixed := make([]int32, m.mixSize)
-
-	for input := range m.inputs {
-		input.mu.Lock()
-
-		if !input.hasBuffered || input.samples.Len() == 0 {
-			input.mu.Unlock()
+func (m *Mixer) mixInputs() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for _, inp := range m.inputs {
+		n, _ := inp.readSample(m.mixTmp[:len(m.mixBuf)])
+		m.mixTmp = m.mixTmp[:n]
```

Comment:
Is data that was late and didn't get mixed skipped anywhere? If not, this may cause sync issues?

Reply:
Is there any buffering logic to allow some packets to arrive late?

Reply:
yeah, I don't see how dejitter can be avoided unless you pass through packets with seqnums intact (i.e., leaving it to the downstream jitter buffer to sort things out). naturally, this only works for cases where mixing is not needed.

Reply:
the general way this sort of thing is handled is via an adaptive jitter buffer as in libwebrtc, where you only incur as much delay as is needed to dejitter most of the packets.

Reply:
@biglittlebigben @juberti Thank you for the valuable feedback! Love it!

Indeed the old and the new mixers do not account for any jitter. As you pointed out, there are no checks for sequence/timestamps.

> Is mixTmp ever grown back to it's full length?

Yes, it's growing to full size when passing to `readSample` and that function returns the final length.

> If not, it looks like the mixer would consume media at a slower rate than it is written?

Even though the answer is "yes", currently the mixer assumes that it should mix at constant rate. Thus, the mix buffer is exactly 1 frame wide, while inputs buffers are larger (5 frames). I think we cannot make it emit frames faster without delaying mixing of other inputs, which I wanted to avoid.

> Even if restoring mixTmp to its full size, jitter in the ticker firing time may mean that the consumption rate may not match the production rate?

That's a very good point and it can also be true temporarily, e.g. if there are no packets, and then input emits N packets in a burst. But the rate should be the same on average. This is why I made input buffer 5x the frame size.

So if the ticker causes jitter, it may lead to one or more inputs to be delayed by 1 frame. Interesting. I didn't consider this. Is this why you made `ignoredLastPrune` variable in the mixer code, right? I'll try to add a test for this.

> Is data that was late and didn't get mixed skipped anywhere? If not, this may cause sync issues?

> Is there any buffering logic to allow some packets to arrive late?

If the packets get delayed, and then arrive late, they will be added to the buffer (max 5 frames), which will effectively delay that particular input relative to others in the mix. But the delay is limited - as soon as we receive more than 5 frames we will start dropping oldest frames/samples from the buffer.

I assume there's no way to really catch up and reduce this delay except dropping frames or trying to speed up and resample audio in oldest frames. Since we are mixing, we cannot really send multiple frames in one go (without delaying everyone else in the mix).

> yeah, I don't see how dejitter can be avoided unless you pass through packets with seqnums intact (i.e., leaving it to the downstream jitter buffer to sort things out). naturally, this only works for cases where mixing is not needed (although that seems like a very important subset).

Indeed, it looks like I should dejitter all the inputs. For now SIP server is a very naive in this regard.

Regarding "no mixing" part, I will also add a pass-through mode later as well (if there's only 1 input in the mix). 

> the general way this sort of thing is handled is via an adaptive jitter buffer as in libwebrtc, where you only incur as much delay as is needed to dejitter most of the packets.

Yeah, looks like it's unavoidable. I'd like to keep the mixer simple, because it shouldn't handle the jitter on its own, right?

I'll then implement a dejitter for each RTP input we receive. So the mixer will receive normalized streams and will just mix those. In that case mixer won't need so much space in input buffers too.

Reply:
I think a mixer has to handle jitter on its own, since it has to emit a single time-aligned stream. this is why mix-less architectures like SFU have become popular.

Reply:
> @biglittlebigben @juberti Thank you for the valuable feedback! Love it!
> 
> Indeed the old and the new mixers do not account for any jitter. As you pointed out, there are no checks for sequence/timestamps.
> 

The new code may not be meaningfully worse than the old one at handling late packets indeed. However, I do believe that we'll need to deal with it eventually. In my pas experience, relying on timestamps has been problematic for RTP track mixing. However, replacing sequence number gaps with silence, and using small leaky queue (with a duration of roughly the duration of the max acceptable sync drift) has worked pretty well. 

Your current approach seems close to this, with a ring buffer that holds up to 100ms of data. However, it looks like you start reading from an input immediately when added without waiting for some buffering level to be reached, which may create stuttering for that track at the beginning.

> > Is mixTmp ever grown back to it's full length?
> 
> Yes, it's growing to full size when passing to `readSample` and that function returns the final length.

Indeed. I missed that.

> 
> > If not, it looks like the mixer would consume media at a slower rate than it is written?
> 
> Even though the answer is "yes", currently the mixer assumes that it should mix at constant rate. Thus, the mix buffer is exactly 1 frame wide, while inputs buffers are larger (5 frames). I think we cannot make it emit frames faster without delaying mixing of other inputs, which I wanted to avoid.
> 
> > Even if restoring mixTmp to its full size, jitter in the ticker firing time may mean that the consumption rate may not match the production rate?
> 
> That's a very good point and it can also be true temporarily, e.g. if there are no packets, and then input emits N packets in a burst. But the rate should be the same on average. I made input buffer 5x the frame size to compensate for this.
> 

I do not see the code path where N packets may be emitted in a burst (the go ticker can drop events https://pkg.go.dev/time#NewTicker). Did I miss it?

> So if the ticker causes jitter, it may lead to one or more inputs to be delayed by 1 frame. Interesting. I didn't consider this can be caused by the timer itself. Is this why you made `ignoredLastPrune` variable in the mixer code, right? I'll try to add a test for this and fix it.
> 
> > Is data that was late and didn't get mixed skipped anywhere? If not, this may cause sync issues?
> 
> > Is there any buffering logic to allow some packets to arrive late?
> 
> If the packets get delayed, and then arrive late, they will be added to the buffer (max 5 frames), which will effectively delay that particular input relative to others in the mix. But the delay is limited - as soon as we receive more than 5 frames we will start dropping oldest frames/samples from the buffer.
> 

ok

> I assume there's no way to really catch up and reduce this delay except dropping frames or trying to speed up and resample audio in oldest frames. Since we are mixing, we cannot really send multiple frames in one go (without delaying everyone else in the mix).
> 
> > yeah, I don't see how dejitter can be avoided unless you pass through packets with seqnums intact (i.e., leaving it to the downstream jitter buffer to sort things out). naturally, this only works for cases where mixing is not needed (although that seems like a very important subset).
> 
> Indeed, it looks like I should dejitter all the inputs. For now SIP server is a very naive in this regard.
> 
> Regarding "no mixing" part, I will add a pass-through mode later as well (if there's only 1 input in the mix).
> 
> > the general way this sort of thing is handled is via an adaptive jitter buffer as in libwebrtc, where you only incur as much delay as is needed to dejitter most of the packets.
> 
> Yeah, looks like it's unavoidable. I'd like to keep the mixer simple, because it shouldn't handle the jitter on its own, right?
> 
> I'll then implement a dejitter for each RTP input stream we receive. So the mixer will receive normalized streams and will just mix those. In that case mixer won't need so much space in input buffers too.

Reply:
> I do not see the code path where N packets may be emitted in a burst (the go ticker can drop events https://pkg.go.dev/time#NewTicker). Did I miss it?

Sorry for confusion, I was referring to the RTP pushing frames to one of the inputs in a burst.

In general it seems like the the change is good enough for now, except two TODO items:

1. Make a proper bypass mode (for 1:1 calls)
2. Start reading only after we get some buffering

For the item (1) I'll need to check what fields in RTP we should keep and which we'll rewrite. We cannot pass packets directly, because we still need OPUS<->PCMU conversion. Also we currently switch mixer tracks on/off when we play pin prompts. That makes it harder to do the bypass. So I'd rather do it in a separate PR.

For the item (2), as @biglittlebigben suggested, we could skip mixing inputs that have less than 2-3 frames (~40-60ms). That will be the constant latency when having multiple participants. Would that be reasonable?

Reply:
I think we should wait until we have 2-3 frames before starting to mix a track, but I think we should consume all available data once running, if needed. 

I also think that we need to address possible dropped tick events, possibly by tracking the delay since the last mix, and running the mixer enough time to emit packets covering the interval.

Reply:
@biglittlebigben Done! Mixer will now wait for 3 frames until mixing the input. It will consume all data if packets stop arriving, at which point it will start buffering again. It will also catch up in case it's scheduled too late.